File: baekjoon/08_implementation/기적의매매법.py
# 20546

N = int(input())
arr = list(map(int, input().split()))

a_m = b_m = N
bnp = 0
timing = 0
for i,a in enumerate(arr):
    while a_m >= a:
        bnp += 1
        a_m -= a

    if i >= 3:
        if arr[i] > arr[i-1] > arr[i-2] > arr[i-3]:
            b_m += timing * a
            timing = 0
        if arr[i] < arr[i-1] < arr[i-2] < arr[i-3]:
            while b_m >= a:
                timing += 1
                b_m -= a

if a_m+bnp*arr[-1] > b_m+timing*arr[-1]:
    print("BNP")
elif a_m+bnp*arr[-1] < b_m+timing*arr[-1]:
    print("TIMING")
else:
    print("SAMESAME")

File: baekjoon/08_implementation/기적의매매법.java
package baekjoon;

import java.util.Scanner;

public class 기적의매매법 {

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);
		int money = sc.nextInt();
		int jun = money;
		int junStock = 0;
		int sung = money;
		int sungStock = 0;
		int[] stock = new int[15];
		for (int i = 1; i <= 14; i++) stock[i] = sc.nextInt();
		for (int i = 1; i <= 14; i++){
			if (jun >= stock[i]) {
				junStock += jun / stock[i];
				jun %= stock[i];
				
			}
		}
		jun += stock[14] * junStock;
		int buycnt = 0;
		int sellcnt = 0;
		for (int i = 1; i <= 14; i++) {
			if (stock[i] < stock[i-1]) {
				buycnt++;
				sellcnt = 0;
			}
			if (stock[i] > stock[i-1]) {
				sellcnt++;
				buycnt = 0;
			}
			if (sellcnt == 3) {
				sung += stock[i] * sungStock;
				if (i > 11) break;
				sungStock = 0;
				sellcnt = 0;
				buycnt = 0;

			}
			if (buycnt >= 3 && sung > stock[i]) {
				sungStock += sung / stock[i];
				sung %= stock[i];
				sellcnt = 0;
				//buycnt = 0;

			}
			if (i == 14) sung += stock[i] * sungStock;
		}
//		System.out.println(jun);
//		System.out.println(sung);
		
		if (sung > jun) System.out.println("TIMING");
		else if (sung < jun) System.out.println("BNP");
		else System.out.println("SAMESAME");
	}

}
